### ISSUE TITLE
`difference_by` would be a nice addition (`anti_join` done)

### ISSUE THREAD

**Issue by jhellerstein** [hydro-project/hydroflow#349]
Often we want the difference of two sets of complex objects, where the matching criterion is custom, and perhaps the rhs type is different from the lhs type.

For example, i may want to output all objects whose id attribute is not in a set of tombstoned ids:
```rust
diff = difference_by() -> for_each(|x| println!("Keeper: {:?}", x));
source_iter(vec!((1,'alice'), (2, 'bob')) -> [pos]diff;
source_iter(vec!(2)) -> map(|i| (i, ())) -> [neg]diff;
```

Current workaround is something like this:
```rust
items = source_iter(vec!((1,'alice'), (2, 'bob')) -> tee();
tombstones = source_iter(vec!(2)) -> tee();

diff_ids = difference();
items -> map(|(i, v)| i) -> [pos]diff;
tombstones -> [neg]diff;

diff = join() -> map(|(i, (v, _))| (i, v)) -> for_each(|x| println!("Keeper: {:?}", x));
items -> [0]diff;
tombstones -> map(|i| (i, ())) -> [1]diff;
```
That's both hard to read and inefficient.

**Comment by MingweiSamuel** [comment]
And maybe `join_by()` #185

**Comment by zzlk** [comment]
shadaj has implemented anti_join in this pr: #382

**Comment by MingweiSamuel** [comment]
#382 solves this issue actually
